### PR TITLE
Grid Layout, Line Chart component, and dynamic generation of charts.

### PR DESCRIPTION
**Pull request by spmcbride1201**
Lots of changes going on, so I'd like to merge into an intermediate `development` branch.

This PR includes code that provides:
* a draggable, resizable grid layout
* a handful of Grey Matter chart components that use a standardized time series format
* A modified JSON inspector that shows the dynamic generation of charts using a standardized Grey Matter chart component.

Targeted next-steps before merging `development` into `master` include:
* the ability to add new chart component to a dashboard
* a standardized Grey Matter dashboard component that gets populated with state via a standardized JSON format
* use of local storage to persist custom dashboards
* [stretch] - The ability to add new dashboards

### CODE REVIEW

File: src/components/HTTPGrid.js [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
 import { connect } from 'react-redux';
-import HTTPConnectionsChart from './HTTPConnectionsChart';
-import DataTxChart from './DataTxChart';
-import HTTPStats from './HTTPStats';
+import GMLineChart from './GMLineChart';
+import GMTable from './GMTable';
 import DataTotals from './DataTotals';
-import { getLatestAttribute, getAttributeOverTime, getAttributeChangesOverTime, mergeResults } from '../utils';
+import { getLatestAttribute, getTimeSeriesOfValue, getTimeSeriesOfNetChange, mergeTimeSeries, mapOverTimeSeries } from '../utils';
+import {Responsive, WidthProvider} from 'react-grid-layout';
```
Comment (tilleryd) [inline src/components/HTTPGrid.js]: add spaces inside the brackets to be consistent

File: src/components/HTTPGrid.js [hunk 2]
```diff
@@ -17,49 +18,78 @@ class HTTPGrid extends Component {
...
+              )}
+              title={`HTTP Connections`}
+            />
+          </div>  
+          <div
+            data-grid={{ x: 4, y: 0, w: 4, h: 6, minW: 3, minH: 4  }}
+            key='b'
+            style={{
+              border: 'solid',
+              overflow: 'hidden'
```
Comment (tilleryd) [inline src/components/HTTPGrid.js]: yeah for sure i think you need a GridCell component...

File: src/components/JSONComponent.js [hunk 3]
```diff
@@ -2,30 +2,66 @@ import React, { Component } from 'react';
 import { PropTypes } from 'prop-types';
 import Inspector from 'react-json-inspector';
 import { connect } from 'react-redux';
+import { extractPaths } from '../utils';
+import GMLineChart from './GMLineChart';
+import { getTimeSeriesOfValue } from '../utils';
+import {Responsive, WidthProvider} from 'react-grid-layout';
```
Comment (tilleryd) [inline src/components/JSONComponent.js]: ditto, spacing inside curly braces

File: src/components/JVMGrid.js [hunk 4]
```diff
@@ -1,9 +1,11 @@
 import React, { Component } from 'react';
 import { PropTypes } from 'prop-types';
 import { connect } from 'react-redux';
-import JVMHeapChart from './JVMHeapChart';
-import JVMClassesChart from './JVMClassesChart';
-import { getLatestAttribute, getAttributeOverTime, mergeResults } from '../utils';
+import GMLineChart from './GMLineChart';
+import { getLatestAttribute, getTimeSeriesOfValue, mergeTimeSeries, mapOverTimeSeries } from '../utils';
+import {Responsive, WidthProvider} from 'react-grid-layout';
```
Comment (tilleryd) [inline src/components/JVMGrid.js]: spacing inside curly braces

File: src/components/RouteGrid.js [hunk 5]
```diff
@@ -1,57 +1,109 @@
...
   };
 
   render() {
-    const { params, route } = this.props;
-    const routeVerbs = route ? Object.keys(route[params.routeName]) : [];
+    const { route } = this.props;
+    const routeName =
+      this.props.match &&
+      this.props.match.params &&
+      this.props.match.params.routeName &&
```
Comment (tilleryd) [inline src/components/RouteGrid.js]: how about `_.hasIn(this.props, ['match', 'params', 'routeName']);`?

File: src/utils.js [hunk 6]
```diff
@@ -33,17 +35,22 @@ export function getLatestAttribute(props, path) {
...
+// TODO: add array at dataAttributes containing all of the keys storing metrics
+// in this time series. This will allow easier lookup
+
 /**
- * Returns the value of an attribute over time as an array of objects with keys of 
- * time(UNIX time in ms), prettyTime(string of time in h:MMtt), and a dynamically generated
- * key (least significant key passed into the path parameter) with the value of the attribute
+ * Returns time series data of a metric's value
+ *   time - UNIX time in ms
+ *   prettyTime - string of time in hMMtt,
```
Comment (tilleryd) [inline src/utils.js]: `hhmmss`?

File: src/utils.js [hunk 7]
```diff
@@ -289,39 +286,124 @@ export function parseJVMMetrics(rawScrapedMetrics) {
...
+      return true;
     }
   });
-  return normalizedTimeseriesData;
 }
+
+/**
+ * @callback dataAttributesMapper
+ * @param {String} currentValue
+ * @returns {String}
```
Comment (tilleryd) [inline src/utils.js]: JSDoc for an invisible function?

File: src/components/FinagleGrid.js [hunk 8]
```diff
@@ -15,36 +14,85 @@ class FinagleGrid extends Component {
...
+          >  
+            <GMBasicMetrics
+              detailLines={{
+                "Initial Resolution": getLatestAttribute(finagle, 'clientregistry.initialresolutionMs'),
+                "Size": getLatestAttribute(finagle, 'futurePool.completedTasks'),
+                "Pool Size": getLatestAttribute(finagle, 'clientregistry.size')
+              }}
+              title="Future Pool"
+            />
+          </div>
```
Comment (spmcbride1201) [inline src/components/FinagleGrid.js]: We failed. Alas, t'was a noble effort.

File: package.json [hunk 9]
```diff
@@ -67,7 +70,6 @@
   "eslintConfig": {
     "extends": "react-app"
   },
-  "proxy": "For JVM, use 'http://localhost:9990'. For Golang, use 'http://localhost:9092'",
```
Comment (spmcbride1201) [inline package.json]: restored!

File: src/utils.js (hunk 7 above)
Comment (spmcbride1201) [inline src/utils.js]: It's supposed to be the expected format of the function that gets passed in as an arg to the following function. First time I've used @callback, so might be jacked up.

**Pull request merged**